flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java

```diff
@@ -62,8 +62,15 @@ public void cleanCheckpoint(
                             }
                         }
                     } finally {
-                        numberOfCheckpointsToClean.decrementAndGet();
-                        postCleanAction.run();
+                        try {
+                            numberOfCheckpointsToClean.decrementAndGet();
+                            postCleanAction.run();
+                        } catch (Exception e) {
+                            LOG.error(
+                                    "Error while cleaning up checkpoint {}",
+                                    checkpoint.getCheckpointID(),
+                                    e);
```

Comment:
I am not entirely sure whether I would swallow the exceptions here. I think the contract should be that here mustn't occur any exceptions. If they do, then this warrants for a hard exit of the process. With a log, even if it is on error, we would probably not have found this bug very soon.

Reply:
> the contract should be that here mustn't occur any exceptions

To guarantee such a contract I see only two options:
1. Wrap `timer.execute` in `CheckpointCoordinator.scheduleTriggerRequest` in a `syncrhonized` block.
2. Tolerate `RejecectionExecutionException && isShutdown` case as I proposed in jira ticket.

The 1st option is deadlock-prone IMO, the 2nd requries some discussion and rework (so not a quick solution).
Logging is also not perfect but to me it's the lesser evil.

Reply:
Hmm, I guess this is related to the threading model of the `CheckpointCoordinator`.

Could you elaborate on the deadlock you are fearing? Which other locks would be hold by the `postCleanAction` which could cause a deadlock?

Reply:
The calling thread already acquires `CheckpointCoordinator.lock` and then `checkpoint.lock`. So it will be enough to synchronize on `checkpoint.lock` in `postCleanAction` to have a deadlock.

Reply:
The calling thread is the thread calling `CheckpointsCleaner.cleanCheckpoint`, right? Then this call will enqueue the clean up action and eventually release the `checkpoint.lock`. That's when the clean up action can complete. 

I think as long as we have made sure that the `CheckpointsCleaner` owns the `Checkpoint` there should be no other thread accessing the `checkpoint.lock`. At the end of the day it probably boils down to a proper lifecycle management of the involved objects and deciding who owns what.

But maybe I am overlooking something here and you have a concrete thread interleaving in mind which I don't see at the moment.

Reply:
No, I don't have a concrete scenario in mind. I'm just concerned that it can arise in the future.
Okay, I'll add `synchronized` here; and drop the other commit - it's not necessary if `postCleanAction` is safe.